File: Difficulty: Medium/Maximum Product Subarray/maximum-product-subarray.cpp
//{ Driver Code Starts
#include <bits/stdc++.h>

using namespace std;


// } Driver Code Ends
// User function Template for C++
class Solution {
  public:
    // Function to find maximum product subarray
    int maxProduct(vector<int> &arr) {
        // Your Code Here
        // 25.11.24 potd
        if(arr.size() == 1) return arr[0];
        int ans=arr[0], n=arr.size(), left=1, right=1;
        
        for(int i=0; i<n; i++){
            if(left == 0)   left = 1;
            if(right == 0)  right = 1;
            
            left *= arr[i];
            right *= arr[n-1-i];
            
            ans = max(ans, max(left,right));
        }
        
        return ans;
    }
};

//{ Driver Code Starts.

int main() {
    int t;
    cin >> t;
    cin.ignore();
    while (t--) {
        // int n, i;
        // cin >> n;
        // vector<int> arr(n);
        // for (i = 0; i < n; i++) {
        //     cin >> arr[i];
        // }

        vector<int> arr;
        string input;

        // Read array
        getline(cin, input);
        stringstream ss(input);
        int number;
        while (ss >> number) {
            arr.push_back(number);
        }
        Solution ob;
        auto ans = ob.maxProduct(arr);
        cout << ans << "\n";

        cout << "~"
             << "\n";
    }
    return 0;
}
// } Driver Code Ends

File: Difficulty: Medium/Maximum Product Subarray/maximum-product-subarray.py
#User function Template for python3
class Solution:

    # Function to find maximum
    # product subarray
    def maxProduct(self,arr):
        # code here
        # 08.02.26 (potd)
        ans, leftProd, rightProd, n = arr[0], 1, 1, len(arr)
        # 'leftProd' for calculating product from left to right
        # 'rightProd' for claculating product from right ot left
        
        for i in range(n):
            # Update leftProd and rightProd if they're 0
            if leftProd == 0:
                leftProd = 1
            if rightProd == 0:
                rightProd = 1
                
            # Multiply respective elements to 'leftProd' and 'rightProd'
            # from left and right end respectively
            leftProd *= arr[i]
            rightProd *= arr[n-1-i]
            
            # Assign max product to 'ans'
            ans = max(ans, max(leftProd, rightProd))
            
        return ans


